app/ts/utils/uniswap.ts

```diff
@@ -0,0 +1,122 @@
+import { AbiCoder, getCreate2Address, keccak256, solidityPacked, Interface } from "ethers"
+import { EthereumAddress } from "./wire-types.js"
+import { addressString } from "./bigint.js"
+import { CHAINS, isSupportedChain } from "./constants.js"
+
+interface UniswapPools {
+	token0IsQuote: boolean
+	v2Pools: EthereumAddress[],
+	v3Pools: EthereumAddress[]
+}
+
+interface Multicall3Call {
+	target: string
+	allowFailure: boolean
+	callData: string
+}
+
+const isSuccessfulCall = ({ returnData, success }: { returnData: string, success: boolean }) => success && returnData !== '0x'
+
+export function calculateUniswapLikePools(token: EthereumAddress, quoteToken: EthereumAddress, chainString: string): UniswapPools | undefined {
+	const [token0, token1] = token < quoteToken ? [addressString(token), addressString(quoteToken)] : [addressString(quoteToken), addressString(token)]
+	const abi = new AbiCoder()
+
+	if (!isSupportedChain(chainString)) return undefined
+
+	const v2Pools = CHAINS[chainString].priceSources.uniswapV2like.map(({ factory, initCodeHash }) => EthereumAddress.parse(getCreate2Address(addressString(factory), keccak256(solidityPacked(['address', 'address'], [token0, token1])), initCodeHash)))
+
+	const v3Pools = CHAINS[chainString].priceSources.uniswapV3Like.map(({ factory, initCodeHash }) => [
+		getCreate2Address(addressString(factory), keccak256(abi.encode(['address', 'address', 'uint24'], [token0, token1, 100])), initCodeHash),
```

Comment:
could create helper function for `getCreate2Address` that would take bigint instead of string